file/src/main/java/akka/stream/alpakka/file/impl/FileTailSource.java

```diff
@@ -111,11 +111,15 @@ public void preStart() {
                 (tryInteger) -> {
                   if (tryInteger.isSuccess()) {
                     int readBytes = tryInteger.get();
+                    // when the number of bytes read is -1 it signals that no new data is available
                     if (readBytes > 0) {
                       buffer.flip();
                       push(out, ByteString.fromByteBuffer(buffer));
                       position += readBytes;
                       buffer.clear();
+                    } else if (!Files.exists(path)) {
```

Comment:
This is a blocking call, I believe it should be wrapped in a future and use an `asyncCallback` to trigger completion.
It should possibly not try this for every poll but have some counter?

Reply:
@ennru Good point.  It will be executed for every poll where no new data was read from the channel, which could be frequent if the log is inactive for some time, but there is still lots of demand from downstream.  I can add the `asyncCallback`, but if we only check every `X` poll then we could have a situation where the stream stays open indefinitely when there's no demand.  That currently happens now as well, but with a counter it would happen more frequently.

This Graph Stage already uses a `TimerGraphStageLogic`.  I could add a 2nd timer which ticks every second to see if the file exists.

Reply:
Another option would be to use `DirectorySourceChange` internally to trigger the shutdown on file deletes, but maybe that's overkill.

Reply:
Schedule a second timer to check if the file is still there when no bytes are read, cancel it when bytes are read.

Reply:
I'll have to add a new parameter to define a new interval for the file check, since reusing the poll interval is effectively check if the file exists as frequently as the current implementation (though, in a non-blocking manner).

Reply:
Maybe that's not so bad though.. I'll reuse the poll duration for now.

Reply:
Trying a different approach with operator DSL.